### ISSUE TITLE
Multiple Classification

### ISSUE THREAD

**Issue by chidiewenike** [MycroftAI/mycroft-precise#24]
If I was to want to consider a multiple classification keyword spotter, what would be the process? I would certainly need to expand the output layer and change to a "softmax" activation function but what about labeling and the back propagation?

**Comment by MatthewScholefield** [comment]
The remaining things left to do would be:
 - **Make the output labels be one-hot vectors instead of ones and zeros:**

To do it properly, this will take a complete restructure of the [TrainData](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/train_data.py) class. As a quicker way to get it working, you can just overwrite [the code to load train and test data here](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/scripts/train.py#L132) (it just needs to return the following tuple of numpy vectors: `((train_in, train_out), (test_in, test_out))`. You would have to write code that lists all the folders in `args.folder` as classes and loads each wav file into the train input vector and also adds a one-hot to the corresponding output vector, [similar to this](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/train_data.py#L196). To load an input vector from a filename do `precise.vectorization.load_vector('myfolder/classname/foo.wav')`.

 - **Handle predictions being class labels**

You have to change the function `precise.network_runner.Runner.run` to return a numpy array of probabilities instead of a single float. Then you have to handle an array [in TriggerDetector](https://github.com/MycroftAI/mycroft-precise/blob/dev/runner/precise_runner/runner.py#L110). There, it might make sense to have `TriggerDetector.update` return the activated class id or `None`, and then receive that class id [here](https://github.com/MycroftAI/mycroft-precise/blob/dev/runner/precise_runner/runner.py#L229) and if it's not `None`, pass it to the `on_activation` handler so that you can receive it [here](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/scripts/listen.py#L51) and print it in the `precise-listen` script.

**Comment by chidiewenike** [comment]
Awesome, I appreciate the quick response to this. I will work on this and report back with my results. I am also looking at porting this to Android so I will report back with my progress on that as well.

**Comment by MatthewScholefield** [comment]
Cool to hear! If you run into any errors and want a faster form of communication, feel free to message me on https://chat.mycroft.ai/ with the username `@matthew-mycroft`.

**Comment by torntrousers** [comment]
Hello @chidiewenike , did  you manage to make any progress on this? I'd like to do the same and it would be great if you'd got somewhere with doing it?

**Comment by stcleezy** [comment]
@MatthewScholefield do you see this getting on the roadmap any time soon?

**Comment by krisgesling** [comment]
Hi @stcleezy, our priority is improving the primary model and adding mechanisms for users to personalize this model for everyone in their household. Unfortunately I don't see multiple wake words being prioritized for our core team in the near future.

**Comment by sushipan** [comment]
Hi, can you give some suggestion about how to change the [ThresholdDecoder](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/threshold_decoder.py#45)  to support multiple wake words ,when use train_incremental 
@MatthewScholefield

**Comment by MatthewScholefield** [comment]
> Hi, can you give some suggestion about how to change the [ThresholdDecoder](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/threshold_decoder.py#45)  to support multiple wake words ,when use train_incremental 
> @MatthewScholefield 

The purpose of the threshold decoder is just to convert squashed probabilities (like 0.999) to more reasonable probabilities (like 0.7) so that 0.7 means around a 70% chance that a wake word was detected.

So basically it's not that important. To just get it working, you can completely remove the ThresholdDecoder. Simply replace `decoder.decode(...)` with `...` (and same with `decoder.encode(...)`.

**Comment by sushipan** [comment]
Thank you , I will try.

**Comment by sushipan** [comment]
@MatthewScholefield Your suggestion help me a lot , but still have some difficult 
Do I need to change  [this](https://github.com/MycroftAI/mycroft-precise/blob/e1a635e9675047eb86f64ca489a1b941321c489a/precise/vectorization.py#L55) np.zero_like  and  [link](https://github.com/MycroftAI/mycroft-precise/blob/e1a635e9675047eb86f64ca489a1b941321c489a/precise/vectorization.py#L78) np.zero  to one-hot ?

**Comment by MatthewScholefield** [comment]
No, the first location you linked can be entirely ignored (delta vectors are not used by default) and the second link is just padding the input.

You want to change the *output* from a single number to a one hot vector. So, right now the training data looks like this (for a bunch of different audio samples):
```python
audio -> [0.0]  # audio with garbage
audio -> [0.0]  # different audio with garbage
audio -> [1.0]  # audio with "hey mycroft"
```

But for multi classification of course we want training data like:
```
audio -> [1.0, 0.0, 0.0]  # garbage
audio -> [0.0, 0.0, 1.0]  # audio with "hey marvin"
audio -> [0.0, 0.0, 1.0]  # different audio with "hey marvin"
audio -> [0.0, 1.0, 0.0]  # audio with "hey mycroft"
```

The part that right now sets the output to `[0.0]` or `[1.0]` is `[output]` [in this specific line in `train_data.py`](https://github.com/MycroftAI/mycroft-precise/blob/dev/precise/train_data.py#L206).

Let me know if this makes sense.

**Comment by sushipan** [comment]
I think this could make sense
